README.md
# Библиотека для роутинга


CHANGELOG.md
### 0.1.1
- Добавлен метод Request::getParams()

### 0.1.0
- Добавлен метод Request::isPost()
- Добавлен метод Request::isGet()
- Добавлен метод Request::getGet()
- Добавлен метод Request::getPost()
- Добавлен метод Request::getParam()

### 0.0.8
- Баг фикс

### 0.0.7
- Рефакторинг

### 0.0.6
- Рефакторинг unit-тестов

### 0.0.5
- Рефакторинг
- Если не указан в роуте method, то не осуществляется проверок метода
- 100% code coverage

### 0.0.4
- Добавлен контроль метода REST

### 0.0.3
- Рефакторинг тестов
- Добавлены placeholder'ы в uri
- Типы placeholder'ов integer(int), string

### 0.0.2
- Исправлен баг с get-параметрами

### 0.0.1
- Построение иерархии yml-маршрутов
- Покрытие unit-тестами
- Router\Exception\NotFound при ненайденном маршруте

tests/Router/RouterTest.php
<?php

namespace Tests\Router;

use Router\Request;
use Router\Router;

class RouterTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Проверка построение объектов запроса
     *
     * @dataProvider providerRoutes
     *
     * @param string $method
     * @param string $uri
     * @param array $params
     */
    public function testResolve_MethodRouteParams_Request(string $method, string $uri, array $params)
    {
        $request = new Request([], [], [], ['REQUEST_METHOD' => $method], []);
        $router = new Router(__DIR__ . '/../../share', 'site', $request);

        $request = $router
            ->setUri($uri)
            ->resolve();

        foreach ($params as $key => $value) {
            $this->assertNotNull($request->getArg($key));
            $this->assertEquals($value, $request->getArg($key));
        }
    }

    /**
     * @return array
     */
    public function providerRoutes()
    {
        return [
            'rootRoute' => [
                'get',
                '/',
                [
                    'param1' => 'test',
                    'param2' => 'test2'
                ]
            ],
            'subRouteRoot' => [
                'get',
                '/test',
                [
                    'uri' => '/',
                    'action' => 'default',
                    'controller' => 'AdminController'
                ]
            ],
            'subRoute' => [
                'get',
                '/test/v2/kkk',
                [
                    'uri' => '/v2/kkk',
                    'action' => 'kk',
                    'controller' => 'AdminController'
                ]
            ],
            'subRouteWithTwoAllowedMethods' => [
                'get',
                '/test/v3/dk',
                [
                    'uri' => '/v3/dk',
                    'action' => 'dk',
                    'controller' => 'AdminController'
                ]
            ],
            'subRouteWithDigitPath' => [
                'get',
                '/sub/34',
                [
                    'uri' => '/34'
                ]
            ],
            'subSubRouteRoot' => [
                'get',
                '/sub/sun',
                [
                    'uri' => '/',
                    'action' => 'default',
                    'controller' => 'AdminController'
                ]
            ],
            'subSubRoute' => [
                'get',
                '/sub/sun/v2/kkk',
                [
                    'uri' => '/v2/kkk',
                    'action' => 'kk',
                    'controller' => 'AdminController'
                ]
            ],
            'subSubRoutePath' => [
                'get',
                '/sub/sun/v3/dk',
                [
                    'uri' => '/v3/dk',
                    'action' => 'dk',
                    'controller' => 'AdminController'
                ]
            ],
            'subSubRouteWithParameters' => [
                'get',
                '/sub/sun/v3/dk?test=1',
                [
                    'uri' => '/v3/dk',
                    'action' => 'dk',
                    'controller' => 'AdminController'
                ]
            ],
            'routeWithPlaceholders' => [
                'put',
                '/test/user/123/edit',
                [
                    'action' => 'updateUser',
                    'controller' => 'UserController',
                    'params' => [
                        'userId' => 123
                    ]
                ]
            ],
            'routeByMethod' => [
                'put',
                '/shop/shop-name',
                [
                    'action' => 'default',
                    'controller' => 'AdminController',
                    'params' => [
                        'shopName' => 'shop-name'
                    ]
                ]
            ],
            'subRouteWithMethodAccess' => [
                'put',
                '/shop/shop-name/user/934/edit',
                [
                    'action' => 'updateUser',
                    'controller' => 'UserController',
                    'params' => [
                        'shopName' => 'shop-name',
                        'userId' => 934
                    ]
                ]
            ],
            'postRouteRoot' => [
                'post',
                '/',
                [
                    'param1' => 'test',
                    'param2' => 'test2'
                ]
            ],
        ];
    }

    /**
     * @dataProvider providerBadUrls
     * @expectedException \Router\Exception\NotFound
     *
     * @param string $method
     * @param string $uri
     */
    public function testResolve_MethodWithUrl_NotFound(string $method, string $uri)
    {
        $request = new Request([], [], [], ['REQUEST_METHOD' => $method], []);
        (new Router(__DIR__ . '/../../share', 'site', $request))
            ->setUri($uri)
            ->resolve();
    }

    /**
     * @return array
     */
    public function providerBadUrls()
    {
        return [
            'subSchemaNotFoundUri' => [
                'get',
                '/test/v4/ts'
            ],
            'notExistsBaseInSubRoute' => [
                'get',
                '/sub',
            ],
            'notExistsInSubSubRoute' => [
                'get',
                '/sub/sun/v4/ts',
            ],
            'fakeRoute' => [
                'get',
                '/jj',
            ],
            'methodNotMatch' => [
                'get',
                '/shop/shop-name/user/934/edit'
            ],
        ];
    }
}


tests/Router/RouterMethodTest.php
<?php declare(strict_types = 1);

namespace Tests\Router;

use Router\Request;

class RouterMethodTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Request method GET
     */
    public function testIsGet_RequestData_Success()
    {
        $request = new Request(
            [],
            [],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_GET
            ],
            []
        );

        $this->assertTrue($request->isGet());
        $this->assertFalse($request->isPost());
    }

    /**
     * Request method POST
     */
    public function testIsPost_RequestData_Success()
    {
        $request = new Request(
            [],
            [],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_POST
            ],
            []
        );

        $this->assertFalse($request->isGet());
        $this->assertTrue($request->isPost());
    }

    /**
     * Request getParam()
     */
    public function testGetGet_Data_Success()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_GET
            ],
            []
        );

        $this->assertEquals('val1', $request->getGet('var1'));
        $this->assertEquals('val2', $request->getPost('var1'));
        $this->assertEquals('val1', $request->getParam('var1'));
        $this->assertEquals('val000', $request->getParam('var2', 'val000'));
        $this->assertEquals('val000', $request->getGet('aaa', 'val000'));
    }

    /**
     * Request getParam()
     */
    public function testGetPost_Data_Success()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_POST
            ],
            []
        );

        $this->assertEquals('val1', $request->getGet('var1'));
        $this->assertEquals('val2', $request->getPost('var1'));
        $this->assertEquals('val2', $request->getParam('var1'));
        $this->assertEquals('val3', $request->getParam('var2', 'val000'));
        $this->assertEquals('val000', $request->getGet('aaa', 'val000'));
    }

    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Method not allowed
     */
    public function testNotExistMethod_Data_InvalidArgumentException()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => 'INVALID_METHOD'
            ],
            []
        );

        $request->getParam('test');
    }

    /**
     * Request GET getParams()
     */
    public function testGetParams_GetRequest_Success()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_GET
            ],
            []
        );

        $this->assertEquals(
            [
                'var1' => 'val1'
            ],
            $request->getParams()
        );
    }

    /**
     * Request POST getParams()
     */
    public function testGetParams_PostRequest_Success()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => Request::METHOD_POST
            ],
            []
        );

        $this->assertEquals(
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            $request->getParams()
        );
    }

    /**
     * Request INVALID METHOD getParams()
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Method not allowed
     */
    public function testGetParams_InvalidRequest_Success()
    {
        $request = new Request(
            [
                'var1' => 'val1',
            ],
            [
                'var1' => 'val2',
                'var2' => 'val3',
            ],
            [],
            [
                'REQUEST_METHOD' => 'INVALID_METHOD'
            ],
            []
        );

        $request->getParams();
    }
}


src/Router/Router.php
<?php declare(strict_types = 1);

namespace Router;

class Router implements RouterInterface
{
    /**
     * @var array
     */
    protected $routes;

    /**
     * @var string
     */
    protected $uri;

    /**
     * @var string
     */
    protected $basePath;

    /**
     * @var string
     */
    protected $schema;

    /**
     * Переменные из placeholder'ов
     * @var array
     */
    protected $vars = [];

    /**
     * Список placeholder'ов
     * @var string[]
     */
    protected $placeholders;

    /**
     * @var RequestInterface
     */
    protected $request;

    protected $replacement = [
        'integer' => '\d+',
        'int' => '\d+',
        'string' => '[\w\d_-]+'
    ];

    /**
     * @var string
     */
    protected $schemaPath;

    public function __construct(string $basePath, string $schema, RequestInterface $request)
    {
        $this->basePath = $basePath;
        $this->schema = $schema;
        $this->request = $request;
        $this->uri = $request->getUri() ?: '/';
    }

    /**
     * @param string $uri
     * @return $this
     */
    public function setUri(string $uri)
    {
        $this->uri = $uri ?: '/';
        return $this;
    }

    /**
     * {@inheritdoc}
     */
    public function setVars(array $vars)
    {
        $this->vars = $vars;
        return $this;
    }

    /**
     * @param string $uri
     * @return string
     */
    protected function filterPlaceholders(string $uri)
    {
        $this->placeholders = [];

        return preg_replace_callback(
            '#\{([\w\d]+):([\w\d]+)\}#si',
            function ($matches) {
                if (isset($this->replacement[$matches[2]])) {
                    $this->placeholders[] = $matches[1];
                    return sprintf('(%s)', $this->replacement[$matches[2]]);
                }

                return $matches[0];
            },
            $uri
        );
    }

    /**
     * Дополнительные проверки маршрута
     * @param array $routeInfo
     * @return bool
     */
    protected function isValid(array $routeInfo): bool
    {
        if (!empty($routeInfo['method'])) {
            return in_array(
                strtoupper($this->request->getMethod()),
                explode(' ', strtoupper($routeInfo['method']))
            );
        }

        return true;
    }

    /**
     * {@inheritdoc}
     */
    public function resolve(): RequestInterface
    {
        $this->uri = preg_replace('/(.*?)\?.*/si', '$1', $this->uri);

        foreach ($this->getRoutes() as $routeInfo) {
            $pattern = $this->filterPlaceholders($routeInfo['uri']);
            $regexp = sprintf('#%s#si', $pattern);

            if (preg_match($regexp, $this->uri, $matches) && $this->isValid($routeInfo)) {
                if (isset($routeInfo['route'])) {
                    array_shift($matches);

                    return $this->getRouterFromInfo($routeInfo)
                        ->setUri(preg_replace($regexp, '', $this->uri))
                        ->setVars(
                            array_merge(
                                $this->vars,
                                array_combine($this->placeholders, $matches)
                            )
                        )
                        ->resolve();
                } elseif (strlen($this->uri) == strlen($matches[0])) {
                    array_shift($matches);
                    $this->setVars(array_merge($this->vars, array_combine($this->placeholders, $matches)));
                    return $this->makeRequest($routeInfo + ['params' => $this->vars]);
                }
            }
        }

        throw new Exception\NotFound();
    }

    /**
     * @param array $params
     * @return RequestInterface
     */
    protected function makeRequest(array $params): RequestInterface
    {
        return $this->request->setArgs($params);
    }

    /**
     * {@inheritdoc}
     */
    public function getRoutes(): array
    {
        if (empty($this->routes)) {
            $this->routes = yaml_parse_file(sprintf('%s/%s.yml', $this->basePath, $this->schema));
        }

        return $this->routes;
    }

    /**
     * Получаем вложенный RouterInterface
     * @param array $routeInfo
     * @return RouterInterface
     */
    protected function getRouterFromInfo(array $routeInfo): RouterInterface
    {
        $class = $routeInfo['class'] ?? static::class;
        return new $class($this->basePath, $routeInfo['route'], $this->request);
    }
}


src/Router/IRouter.php
<?php

namespace Router;

interface IRouter
{
    /**
     * Uri запроса для которого ищется подходящий сценарий
     *
     * @param string $uri
     *
     * @return $this
     */
    public function setUri(string $uri);

    /**
     * Шаблон для поиска по запросу
     *
     * @param string $patternField
     *
     * @return $this
     */
    public function setPatternField(string $patternField);

    /**
     * Схема
     *
     * @param string $schema
     *
     * @return $this
     */
    public function setSchema(string $schema);

    /**
     * Папка в которой ищется схема
     *
     * @param string $basePath
     *
     * @return $this
     */
    public function setBasePath(string $basePath);

    /**
     * Возвращает подходящий сценарий в виде объекта IRequest
     *
     * @return IRequest
     *
     * @throws Exception\NotFound
     */
    public function resolve();

    /**
     * Список маршрутов
     *
     * @return array
     */
    public function getRoutes();

    /**
     * Переменные
     *
     * @param array $vars
     *
     * @return $this
     */
    public function setVars(array $vars);

    /**
     * REST - Метод
     *
     * @param string $method
     *
     * @return $this
     */
    public function setMethod(string $method);
}


src/Router/Request.php
<?php declare(strict_types = 1);

namespace Router;

class Request implements RequestInterface
{
    const METHOD_POST = 'POST';
    const METHOD_GET = 'GET';
    const METHOD_PUT = 'PUT';
    const METHOD_DELETE = 'DELETE';
    const METHOD_PATCH = 'PATCH';

    /**
     * @var array
     */
    private $get;

    /**
     * @var array
     */
    private $post;

    /**
     * @var array
     */
    private $cookie;

    /**
     * @var array
     */
    private $server;

    /**
     * @var array
     */
    private $files;

    /**
     * @var array
     */
    private $args;

    protected static $instance;

    public function __construct(array $get, array $post, array $cookie, array $server, array $files)
    {
        $this->get = $get;
        $this->post = $post;
        $this->cookie = $cookie;
        $this->server = $server;
        $this->files = $files;
    }

    /**
     * @return RequestInterface
     */
    public static function instance(): RequestInterface
    {
        if (!static::$instance) {
            static::$instance = new static($_GET, $_POST, $_COOKIE, $_SERVER, $_FILES);
        }
        
        return static::$instance;
    }

    /**
     * @param array $args
     * @return $this
     */
    public function setArgs(array $args)
    {
        $this->args = $args;
        return $this;
    }

    /**
     * @param string $name
     * @return mixed|null
     */
    public function getArg(string $name)
    {
        return $this->args[$name] ?? null;
    }

    /**
     * @return null|string
     */
    public function getUri(): ?string
    {
        return $this->server['REQUEST_URI'] ?? null;
    }

    /**
     * @return string
     */
    public function getMethod(): string
    {
        return $this->server['REQUEST_METHOD'];
    }

    /**
     * @return bool
     */
    public function isPost(): bool
    {
        return $this->server['REQUEST_METHOD'] == self::METHOD_POST;
    }

    /**
     * @return bool
     */
    public function isGet(): bool
    {
        return $this->server['REQUEST_METHOD'] == self::METHOD_GET;
    }

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     */
    public function getGet(string $paramName, $defaultValue = null)
    {
        return $this->get[$paramName] ?? $defaultValue;
    }

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     */
    public function getPost(string $paramName, $defaultValue = null)
    {
        return $this->post[$paramName] ?? $defaultValue;
    }

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     * @throws \InvalidArgumentException
     */
    public function getParam(string $paramName, $defaultValue = null)
    {
        switch ($this->getMethod()) {
            case self::METHOD_GET:
                return $this->getGet($paramName, $defaultValue);
            case self::METHOD_POST:
                return $this->getPost($paramName, $defaultValue);
        }

        throw new \InvalidArgumentException('Method not allowed');
    }

    /**
     * @return array
     * @throws \InvalidArgumentException
     */
    public function getParams(): array
    {
        switch ($this->getMethod()) {
            case self::METHOD_GET:
                return $this->get;
            case self::METHOD_POST:
                return $this->post;
        }

        throw new \InvalidArgumentException('Method not allowed');
    }
}


src/Router/RequestInterface.php
<?php declare(strict_types = 1);

namespace Router;

interface RequestInterface
{
    /**
     * @return null|string
     */
    public function getUri(): ?string;

    /**
     * @return string
     */
    public function getMethod(): string;

    /**
     * @param array $args
     * @return $this
     */
    public function setArgs(array $args);

    /**
     * @param string $name
     * @return null
     */
    public function getArg(string $name);

    /**
     * @return bool
     */
    public function isPost(): bool;

    /**
     * @return bool
     */
    public function isGet(): bool;

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     */
    public function getGet(string $paramName, $defaultValue = null);

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     */
    public function getPost(string $paramName, $defaultValue = null);

    /**
     * @param string $paramName
     * @param mixed|null $defaultValue
     * @return mixed|null
     * @throws \InvalidArgumentException
     */
    public function getParam(string $paramName, $defaultValue = null);

    /**
     * @return array
     * @throws \InvalidArgumentException
     */
    public function getParams(): array;
}


share/fixtures/routes.php
<?php

return [
    // #0
    [
        'get',
        '/',
        [
            'param1' => 'test',
            'param2' => 'test2'
        ]
    ],
    // #1
    [
        'get',
        '/test',
        [
            'uri' => '/',
            'action' => 'default',
            'controller' => 'AdminController'
        ]
    ],
    // #2
    [
        'get',
        '/test/v2/kkk',
        [
            'uri' => '/v2/kkk',
            'action' => 'kk',
            'controller' => 'AdminController'
        ]
    ],
    // #3
    [
        'get',
        '/test/v3/dk',
        [
            'uri' => '/v3/dk',
            'action' => 'dk',
            'controller' => 'AdminController'
        ]
    ],
    // #6
    [
        'get',
        '/sub/34',
        [
            'uri' => '/34'
        ]
    ],
    // #7
    [
        'get',
        '/sub/sun',
        [
            'uri' => '/',
            'action' => 'default',
            'controller' => 'AdminController'
        ]
    ],
    // #8
    [
        'get',
        '/sub/sun/v2/kkk',
        [
            'uri' => '/v2/kkk',
            'action' => 'kk',
            'controller' => 'AdminController'
        ]
    ],
    // #9
    [
        'get',
        '/sub/sun/v3/dk',
        [
            'uri' => '/v3/dk',
            'action' => 'dk',
            'controller' => 'AdminController'
        ]
    ],
    // #12
    [
        'get',
        '/sub/sun/v3/dk?test=1',
        [
            'uri' => '/v3/dk',
            'action' => 'dk',
            'controller' => 'AdminController'
        ]
    ],
    // #13
    [
        'put',
        '/test/user/123/edit',
        [
            'action' => 'updateUser',
            'controller' => 'UserController',
            'params' => [
                'userId' => 123
            ]
        ]
    ],
    // #14
    [
        'put',
        '/shop/shop-name',
        [
            'action' => 'default',
            'controller' => 'AdminController',
            'params' => [
                'shopName' => 'shop-name'
            ]
        ]
    ],
    // #15
    [
        'put',
        '/shop/shop-name/user/934/edit',
        [
            'action' => 'updateUser',
            'controller' => 'UserController',
            'params' => [
                'shopName' => 'shop-name',
                'userId' => 934
            ]
        ]
    ],
    // #16
    [
        'post',
        '/',
        [
            'param1' => 'test',
            'param2' => 'test2'
        ]
    ],
];
